polkadot/node/network/availability-recovery/src/task.rs

```diff
@@ -0,0 +1,851 @@
+		&mut self,
+		params: &RecoveryParams,
+		sender: &mut Sender,
+	) -> Vec<ValidatorIndex> {
+		let (tx, rx) = oneshot::channel();
+		sender
+			.send_message(AvailabilityStoreMessage::QueryAllChunks(params.candidate_hash, tx))
+			.await;
+
+		match rx.await {
+			Ok(chunks) => {
+				// This should either be length 1 or 0. If we had the whole data,
+				// we wouldn't have reached this stage.
+				let chunk_indices: Vec<_> = chunks.iter().map(|c| c.index).collect();
+
+				for chunk in chunks {
+					if is_chunk_valid(params, &chunk) {
+						gum::trace!(
+							target: LOG_TARGET,
+							candidate_hash = ?params.candidate_hash,
+							validator_index = ?chunk.index,
+							"Found valid chunk on disk"
+						);
+						self.insert_chunk(chunk.index, chunk);
+					} else {
+						gum::error!(
+							target: LOG_TARGET,
+							"Loaded invalid chunk from disk! Disk/Db corruption _very_ likely - please fix ASAP!"
+						);
+					};
+				}
+
+				chunk_indices
+			},
+			Err(oneshot::Canceled) => {
+				gum::warn!(
+					target: LOG_TARGET,
+					candidate_hash = ?params.candidate_hash,
+					"Failed to reach the availability store"
+				);
+
+				vec![]
+			},
+		}
+	}
+
+	/// Launch chunk requests in parallel, according to the parameters.
+	async fn launch_parallel_chunk_requests<Sender>(
+		&mut self,
+		params: &RecoveryParams,
+		sender: &mut Sender,
+		desired_requests_count: usize,
+		validators: &mut VecDeque<ValidatorIndex>,
+	) where
+		Sender: overseer::AvailabilityRecoverySenderTrait,
+	{
+		let candidate_hash = &params.candidate_hash;
+		let already_requesting_count = self.requesting_chunks.len();
+
+		let mut requests = Vec::with_capacity(desired_requests_count - already_requesting_count);
+
+		while self.requesting_chunks.len() < desired_requests_count {
+			if let Some(validator_index) = validators.pop_back() {
+				let validator = params.validator_authority_keys[validator_index.0 as usize].clone();
+				gum::trace!(
+					target: LOG_TARGET,
+					?validator,
+					?validator_index,
+					?candidate_hash,
+					"Requesting chunk",
+				);
+
+				// Request data.
+				let raw_request = req_res::v1::ChunkFetchingRequest {
+					candidate_hash: params.candidate_hash,
+					index: validator_index,
+				};
+
+				let (req, res) = OutgoingRequest::new(Recipient::Authority(validator), raw_request);
+				requests.push(Requests::ChunkFetchingV1(req));
+
+				params.metrics.on_chunk_request_issued();
+				let timer = params.metrics.time_chunk_request();
+
+				self.requesting_chunks.push(Box::pin(async move {
+					let _timer = timer;
+					match res.await {
+						Ok(req_res::v1::ChunkFetchingResponse::Chunk(chunk)) =>
+							Ok(Some(chunk.recombine_into_chunk(&raw_request))),
+						Ok(req_res::v1::ChunkFetchingResponse::NoSuchChunk) => Ok(None),
+						Err(e) => Err((validator_index, e)),
+					}
+				}));
+			} else {
+				break
+			}
+		}
+
+		sender
```

Comment:
I think we can improve the fetching a bit here. We could skip initiating a new connection to avoid the overhead, or a possible offline peer and initially try to fetch from peers we are already connected. If that doesn't yield enough chunks, we can `TryConnect`.

Reply:
that's an interesting idea!

I focused in this PR on maintaining backwards-compatibility with the previous implementation, in order to avoid introducing bugs and to keep changes reviewable.

I'll include this in a follow-up PR with optimisations, sounds good?

Reply:
We should already be connected to all validators. `TryConnect` should not be needed and likely only adds latency without any benefit. (We removed `TryConnect` in the past because of this.

In other words, the stack already tries to have established connections, if that failed it is unlikely that `TryConnect` will succeed.

Reply:
So you suggest that I remove `TryConnect` from all strategies?
I see that availability-recovery is the only subsystem that uses it

Reply:
I think we could do that for the regular chunk requesting strategy, but we may want to still do a `TryConnect` for systematic chunks recovery (which is coming in a follow-up PR), because a failure to retrieve a single chunk will mess up the entire strategy.